File: project-{topic}-plus/src/controllers/task_controller.cpp
#include "task_controller.h"

TaskController::TaskController() {
    LOG_INFO("TaskController initialized.");
}

Pistache::Rest::RouteCallback TaskController::get_all_tasks() {
    return Pistache::Rest::Routes::Get([&](const Pistache::Rest::Request& request, Pistache::Http::ResponseWriter response) {
        AuthContext* auth_context = static_cast<AuthContext*>(request.getUserData(reinterpret_cast<void*>(1)));
        if (!auth_context) {
            throw InternalServerException("Authentication context missing for /tasks endpoint.");
        }

        try {
            std::vector<Task> tasks;
            if (auth_context->role == UserRole::ADMIN) {
                // Admins can see all tasks
                tasks = Task::find_all();
                LOG_INFO("Admin user " + auth_context->username + " fetched all tasks.");
            } else {
                // Regular users only see their own tasks
                tasks = Task::find_by_user_id(auth_context->user_id);
                LOG_INFO("User " + auth_context->username + " fetched their own tasks.");
            }

            Json::Value tasks_json = Json::arrayValue;
            for (const auto& task : tasks) {
                tasks_json.append(task.to_json());
            }

            Json::Value response_json;
            response_json["status"] = "success";
            response_json["data"] = tasks_json;

            response.headers().add<Pistache::Http::Header::ContentType>(Pistache::Http::Mime::MediaType("application/json"));
            response.send(Pistache::Http::Code::Ok, JsonUtil::to_string(response_json)).get();
        } catch (const ApiException& e) {
            throw;
        } catch (const std::exception& e) {
            LOG_ERROR("Error fetching tasks for user " + auth_context->username + ": " + std::string(e.what()));
            throw InternalServerException("Failed to retrieve tasks.");
        }
    });
}

Pistache::Rest::RouteCallback TaskController::get_task_by_id() {
    return Pistache::Rest::Routes::Get([&](const Pistache::Rest::Request& request, Pistache::Http::ResponseWriter response) {
        AuthContext* auth_context = static_cast<AuthContext*>(request.getUserData(reinterpret_cast<void*>(1)));
        if (!auth_context) {
            throw InternalServerException("Authentication context missing for /tasks/:id endpoint.");
        }

        long task_id;
        try {
            task_id = std::stol(request.param(":id").as<std::string>());
        } catch (const std::exception&) {
            throw BadRequestException("Invalid task ID format.");
        }

        try {
            // Try to get from cache first
            if (std::optional<Task> cached_task = task_cache_.get("task_" + std::to_string(task_id))) {
                 if (auth_context->role == UserRole::ADMIN || cached_task->user_id == auth_context->user_id) {
                    Json::Value response_json;
                    response_json["status"] = "success";
                    response_json["data"] = cached_task->to_json();
                    LOG_DEBUG("Task ID " + std::to_string(task_id) + " fetched from cache.");
                    response.headers().add<Pistache::Http::Header::ContentType>(Pistache::Http::Mime::MediaType("application/json"));
                    response.send(Pistache::Http::Code::Ok, JsonUtil::to_string(response_json)).get();
                    return;
                }
            }

            std::optional<Task> task = Task::find_by_id(task_id);

            if (!task) {
                throw NotFoundException("Task with ID " + std::to_string(task_id) + " not found.");
            }

            // Authorization check: Only owner or admin can view a task
            if (auth_context->role != UserRole::ADMIN && task->user_id != auth_context->user_id) {
                throw ForbiddenException("Access denied. You do not own this task.");
            }

            // Cache the task
            task_cache_.set("task_" + std::to_string(task_id), *task, 60); // Cache for 60 seconds
            LOG_DEBUG("Task ID " + std::to_string(task_id) + " fetched from DB and cached.");


            Json::Value response_json;
            response_json["status"] = "success";
            response_json["data"] = task->to_json();

            response.headers().add<Pistache::Http::Header::ContentType>(Pistache::Http::Mime::MediaType("application/json"));
            response.send(Pistache::Http::Code::Ok, JsonUtil::to_string(response_json)).get();
        } catch (const ApiException& e) {
            throw;
        } catch (const std::exception& e) {
            LOG_ERROR("Error fetching task ID " + std::to_string(task_id) + ": " + std::string(e.what()));
            throw InternalServerException("Failed to retrieve task.");
        }
    });
}

Pistache::Rest::RouteCallback TaskController::create_task() {
    return Pistache::Rest::Routes::Post([&](const Pistache::Rest::Request& request, Pistache::Http::ResponseWriter response) {
        AuthContext* auth_context = static_cast<AuthContext*>(request.getUserData(reinterpret_cast<void*>(1)));
        if (!auth_context) {
            throw InternalServerException("Authentication context missing for /tasks endpoint.");
        }

        std::optional<Json::Value> json_body = JsonUtil::parse_json(request.body());
        if (!json_body) {
            throw BadRequestException("Invalid JSON in request body.");
        }

        if (!JsonUtil::has_required_fields(*json_body, {"title", "description", "due_date"})) {
            throw BadRequestException("Missing required fields: title, description, due_date.");
        }

        std::string title = JsonUtil::get_string(*json_body, "title");
        std::string description = JsonUtil::get_string(*json_body, "description");
        std::string status_str = JsonUtil::get_string(*json_body, "status", "pending");
        std::string due_date = JsonUtil::get_string(*json_body, "due_date");

        try {
            TaskStatus status = string_to_task_status(status_str);
            // Task is created by the authenticated user
            std::optional<Task> new_task = Task::create(title, description, status, due_date, auth_context->user_id);

            if (new_task) {
                // Invalidate cache for all tasks or specific user tasks if such caching was implemented
                // For simplicity, for single item, we just don't add it to cache here,
                // it will be cached on first GET request.

                Json::Value response_json;
                response_json["status"] = "success";
                response_json["message"] = "Task created successfully.";
                response_json["task"] = new_task->to_json();

                response.headers().add<Pistache::Http::Header::ContentType>(Pistache::Http::Mime::MediaType("application/json"));
                response.send(Pistache::Http::Code::Created, JsonUtil::to_string(response_json)).get();
            } else {
                throw InternalServerException("Failed to create task due to an unknown error.");
            }
        } catch (const ApiException& e) {
            throw;
        } catch (const std::exception& e) {
            LOG_ERROR("Error creating task for user " + auth_context->username + ": " + std::string(e.what()));
            throw InternalServerException("Failed to create task.");
        }
    });
}

Pistache::Rest::RouteCallback TaskController::update_task() {
    return Pistache::Rest::Routes::Put([&](const Pistache::Rest::Request& request, Pistache::Http::ResponseWriter response) {
        AuthContext* auth_context = static_cast<AuthContext*>(request.getUserData(reinterpret_cast<void*>(1)));
        if (!auth_context) {
            throw InternalServerException("Authentication context missing for /tasks/:id endpoint.");
        }

        long task_id;
        try {
            task_id = std::stol(request.param(":id").as<std::string>());
        } catch (const std::exception&) {
            throw BadRequestException("Invalid task ID format.");
        }

        std::optional<Json::Value> json_body = JsonUtil::parse_json(request.body());
        if (!json_body) {
            throw BadRequestException("Invalid JSON in request body.");
        }

        try {
            std::optional<Task> existing_task = Task::find_by_id(task_id);
            if (!existing_task) {
                throw NotFoundException("Task with ID " + std::to_string(task_id) + " not found.");
            }

            // Authorization check: Only owner or admin can update a task
            if (auth_context->role != UserRole::ADMIN && existing_task->user_id != auth_context->user_id) {
                throw ForbiddenException("Access denied. You do not own this task.");
            }

            // Update fields if present in JSON body
            if (json_body->isMember("title")) {
                existing_task->title = JsonUtil::get_string(*json_body, "title");
            }
            if (json_body->isMember("description")) {
                existing_task->description = JsonUtil::get_string(*json_body, "description");
            }
            if (json_body->isMember("status")) {
                existing_task->status = string_to_task_status(JsonUtil::get_string(*json_body, "status"));
            }
            if (json_body->isMember("due_date")) {
                existing_task->due_date = JsonUtil::get_string(*json_body, "due_date");
            }

            if (existing_task->update()) {
                // Invalidate cache for this specific task
                task_cache_.remove("task_" + std::to_string(task_id));
                LOG_DEBUG("Task ID " + std::to_string(task_id) + " updated and cache invalidated.");

                Json::Value response_json;
                response_json["status"] = "success";
                response_json["message"] = "Task updated successfully.";
                response_json["task"] = existing_task->to_json();

                response.headers().add<Pistache::Http::Header::ContentType>(Pistache::Http::Mime::MediaType("application/json"));
                response.send(Pistache::Http::Code::Ok, JsonUtil::to_string(response_json)).get();
            } else {
                throw InternalServerException("Failed to update task with ID " + std::to_string(task_id) + ".");
            }
        } catch (const ApiException& e) {
            throw;
        } catch (const std::exception& e) {
            LOG_ERROR("Error updating task ID " + std::to_string(task_id) + ": " + std::string(e.what()));
            throw InternalServerException("Failed to update task.");
        }
    });
}

Pistache::Rest::RouteCallback TaskController::delete_task() {
    return Pistache::Rest::Routes::Delete([&](const Pistache::Rest::Request& request, Pistache::Http::ResponseWriter response) {
        AuthContext* auth_context = static_cast<AuthContext*>(request.getUserData(reinterpret_cast<void*>(1)));
        if (!auth_context) {
            throw InternalServerException("Authentication context missing for /tasks/:id endpoint.");
        }
        
        // Role check (Admin only) is handled by JwtMiddleware::require_role in server.cpp before this handler is called.
        // So, we just proceed with the deletion logic here.

        long task_id;
        try {
            task_id = std::stol(request.param(":id").as<std::string>());
        } catch (const std::exception&) {
            throw BadRequestException("Invalid task ID format.");
        }

        try {
            std::optional<Task> existing_task = Task::find_by_id(task_id);
            if (!existing_task) {
                throw NotFoundException("Task with ID " + std::to_string(task_id) + " not found.");
            }

            // Although `require_role(ADMIN)` is before this,
            // a non-admin could try to delete their own task.
            // If we wanted to allow users to delete their own tasks AND admins to delete any,
            // the logic here would be: `if (auth_context->role != UserRole::ADMIN && existing_task->user_id != auth_context->user_id)`
            // But since the route is configured to `require_role(ADMIN)`, only admins reach this point.
            // No explicit user_id check is needed here due to the `require_role(ADMIN)` middleware.

            if (existing_task->remove()) {
                // Invalidate cache for this specific task
                task_cache_.remove("task_" + std::to_string(task_id));
                LOG_DEBUG("Task ID " + std::to_string(task_id) + " deleted and cache invalidated.");

                Json::Value response_json;
                response_json["status"] = "success";
                response_json["message"] = "Task deleted successfully.";

                response.headers().add<Pistache::Http::Header::ContentType>(Pistache::Http::Mime::MediaType("application/json"));
                response.send(Pistache::Http::Code::Ok, JsonUtil::to_string(response_json)).get();
            } else {
                throw InternalServerException("Failed to delete task with ID " + std::to_string(task_id) + ".");
            }
        } catch (const ApiException& e) {
            throw;
        } catch (const std::exception& e) {
            LOG_ERROR("Error deleting task ID " + std::to_string(task_id) + ": " + std::string(e.what()));
            throw InternalServerException("Failed to delete task.");
        }
    });
}
```

File: project-{topic}-plus/backend/src/modules/tasks/task.controller.ts
import { Request, Response, NextFunction } from 'express';
import { StatusCodes } => from 'http-status-codes';
import * as taskService from './task.service';
import * as projectService from '../projects/project.service'; // To check project ownership
import { AppError } from '../../utils/appError';
import { Role } from '@prisma/client';

/**
 * Handles creation of a new task.
 * Only accessible by ADMIN or the Project Manager of the associated project.
 */
export const createTaskHandler = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const taskData = req.body;
    const { projectId } = taskData;

    if (!req.user) {
        return next(new AppError('Authentication required to create task.', StatusCodes.UNAUTHORIZED));
    }

    const project = await projectService.getProjectById(projectId);
    if (!project) {
        return next(new AppError('Project not found.', StatusCodes.NOT_FOUND));
    }

    // Only ADMIN or the project's manager can create tasks in this project
    if (req.user.role !== Role.ADMIN && req.user.id !== project.managerId) {
      return next(new AppError('You do not have permission to create tasks in this project.', StatusCodes.FORBIDDEN));
    }

    const newTask = await taskService.createTask(taskData);
    res.status(StatusCodes.CREATED).json({
      status: 'success',
      data: { task: newTask },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * Handles retrieval of all tasks.
 * Access depends on user role and query parameters.
 */
export const getAllTasksHandler = async (req: Request, res: Response, next: NextFunction) => {
  try {
    if (!req.user) {
      return next(new AppError('Authentication required to access tasks.', StatusCodes.UNAUTHORIZED));
    }
    const tasks = await taskService.getAllTasks(req.query as any, req.user.id, req.user.role);
    res.status(StatusCodes.OK).json({
      status: 'success',
      results: tasks.length,
      data: { tasks },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * Handles retrieval of a single task by ID.
 * Access depends on user role and association with the task/project.
 */
export const getTaskByIdHandler = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params;
    const task = await taskService.getTaskById(id);

    if (!task) {
      return next(new AppError('Task not found', StatusCodes.NOT_FOUND));
    }

    if (!req.user) {
      return next(new AppError('Authentication required to view task.', StatusCodes.UNAUTHORIZED));
    }

    // ADMIN, PROJECT_MANAGER (of the project), or ASSIGNED_MEMBER can view
    const isProjectManager = task.project?.managerId === req.user.id;
    const isAssignedToUser = task.assignedToId === req.user.id;

    if (req.user.role === Role.ADMIN || isProjectManager || (req.user.role === Role.MEMBER && isAssignedToUser)) {
      return res.status(StatusCodes.OK).json({
        status: 'success',
        data: { task },
      });
    }

    return next(new AppError('You do not have permission to view this task.', StatusCodes.FORBIDDEN));
  } catch (error) {
    next(error);
  }
};

/**
 * Handles updating an existing task.
 * Access depends on user role and association.
 * MEMBER can only update their own task's status.
 * PM can update any task in their project.
 * ADMIN can update any task.
 */
export const updateTaskHandler = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params;
    const updateData = req.body;

    if (!req.user) {
      return next(new AppError('Authentication required to update task.', StatusCodes.UNAUTHORIZED));
    }

    const existingTask = await taskService.getTaskById(id);
    if (!existingTask) {
        return next(new AppError('Task not found', StatusCodes.NOT_FOUND));
    }

    const isProjectManager = existingTask.project?.managerId === req.user.id;
    const isAssignedToUser = existingTask.assignedToId === req.user.id;

    if (req.user.role === Role.ADMIN) {
      // ADMIN can update anything
    } else if (req.user.role === Role.PROJECT_MANAGER && isProjectManager) {
      // PM can update any fields for tasks in their project
    } else if (req.user.role === Role.MEMBER && isAssignedToUser) {
      // MEMBER can only update 'status' of their assigned tasks
      if (Object.keys(updateData).length === 1 && updateData.status) {
        // Only allow status update
        req.body = { status: updateData.status }; // Ensure only status is passed to service
      } else {
        return next(new AppError('Members can only update the status of their assigned tasks.', StatusCodes.FORBIDDEN));
      }
    } else {
      return next(new AppError('You do not have permission to update this task.', StatusCodes.FORBIDDEN));
    }

    const updatedTask = await taskService.updateTask(id, updateData);
    res.status(StatusCodes.OK).json({
      status: 'success',
      data: { task: updatedTask },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * Handles deleting a task.
 * Only accessible by ADMIN or the Project Manager of the associated project.
 */
export const deleteTaskHandler = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params;

    if (!req.user) {
      return next(new AppError('Authentication required to delete task.', StatusCodes.UNAUTHORIZED));
    }

    const existingTask = await taskService.getTaskById(id);
    if (!existingTask) {
        return next(new AppError('Task not found', StatusCodes.NOT_FOUND));
    }

    // Only ADMIN or the project's manager can delete tasks in this project
    const isProjectManager = existingTask.project?.managerId === req.user.id;
    if (req.user.role !== Role.ADMIN && !isProjectManager) {
      return next(new AppError('You do not have permission to delete this task.', StatusCodes.FORBIDDEN));
    }

    await taskService.deleteTask(id);
    res.status(StatusCodes.NO_CONTENT).json({
      status: 'success',
      data: null,
    });
  } catch (error) {
    next(error);
  }
};
```